### ISSUE TITLE
Automate using class rosters for resource differentiation

### ISSUE THREAD

**Issue by yuvipanda** [berkeley-dsep-infra/datahub#476]
Right now, we maintain a secret list of student names -> class name mappings for classes that need more resources than usual. This is error prone & manual, and we should automate this.

We should have:

1. A python script that can talk to grouper to get a list of all student names in our list of classes, and write them out as YAML mapping student names to classes they are in. This should refresh every few minutes, if grouper can support that load.
2. Logic in https://github.com/berkeley-dsep-infra/datahub/blob/staging/hub/values.yaml#L27 that can read this generated YAML file, and determine the resources offered to the student based on the highest ask in all the classes they are enrolled in

We can then run this script as a sidecar next to the hub pod so the data is up to date.

**Comment by ericvd-ucb** [comment]
LS88 Connector  Sarah Reynolds has asked for extra memory - as used in the past.  My understanding is that she sends a roster for manually adding to some sort of whitelist? @ryanlovett ?

**Comment by ryanlovett** [comment]
@yuvipanda An alternative idea is to automatically provision the CalGroup to Google and then use the group scope during OAuth. That way we don't have to run our own script, just define the groups we accept. We would have to add support for this to the authenticator though.

**Comment by yuvipanda** [comment]
@ryanlovett how hard is it to 'automatically provision the CalGroup to Google'?

**Comment by ryanlovett** [comment]
@yuvipanda I've just uploaded https://github.com/ryanlovett/sis2calgroups. I need to add error handling, documentation, automatic provisioning to Google Groups, and a composite group for each course.

I've successfully used it to create CalGroups for Data 8, Data 100, a connector, and another course. If you want to see, try logging into and browsing around calgroups.berkeley.edu.

**Comment by ryanlovett** [comment]
I decided to break apart sis2calgroups into several components. I need to:
 - [x] write a [grouper cli](https://github.com/ryanlovett/grouper-cli) tool to manipulate CalGroups
 - [ ] write a corresponding sis cli command
 - [ ] run a sis-to-calgroup workflow in some sort of CI manner
 - [ ] add group scope support to the Google oauthenticator
 - [ ] reference group scope for pod limits

**Comment by ryanlovett** [comment]
@yuvipanda Should the user be enrolled in more than one course, my feeling is that they should _not_ get the max of the resource limits specified for any of their courses. This would unfair to their peers since the pedagogy might require that students accomplish tasks with various constraints. I think this means that we must present a profile choice for students who are enrolled in more than one course where at least one of those courses has defined resource limits, singleuser images, or mounts. Alternatively we present this choice for students who are enrolled in more than one course, regardless of whether any resources have been customized.

Some students might choose to do work for one course under the profile for another. If this is a problem, assignment submissions could pass in the name of the course profile as an attribute.

What do you think?

**Comment by yuvipanda** [comment]
@ryanlovett I think asking students to explicitly choose profiles should be avoided at all costs, at least to begin with. We can experiment with profiles first over the next semester, and see where we go after that. I think students enrolled in these courses get extra RAM right now, and we should preserve that behavior - just automate it. We can do more per-class limitations later, *if* they are pedagogically important. I suspect they aren't - students can always do it in their local computers with more RAM, for example. But we should investigate that as a separate setup, and not as part of this.

**Comment by ryanlovett** [comment]
Okay. To your point that this is happening right now I think we should make sure the instructors are aware of it, just to be safe. Similar to the admin situation.

**Comment by ericvd-ucb** [comment]
@ryanlovett I think its ok to give them the Max Ram of two possible profiles. I don't think its a common issue that instructors expect users to complete work under a certain memory limit. I don't think that we should make users choose profiles. I dont see RAM limiting as a pedagogically important concept. Open to being proved wrong of course.

**Comment by ryanlovett** [comment]
Thanks @ericvd-ucb! I think I've over estimated how important it is to set identical limits. Still, I think some faculty may want to be aware that not all of their students have the same constraints on datahub (even if they already know not all have the same personal computing resources).

**Comment by ryanlovett** [comment]
@yuvipanda Can you comment on this strategy?

hub has config resembling:
```
    extraConfigMap:
      profiles:
        2019-summer-12345: # year-term-classid
          mem_limit: 4096M
          mem_guarantee: 4096M
        2019-fall-23456:
          mem_limit: 2048M
          mem_guarantee: 2048M
          image: # e.g., but not for this
            name: gcr.io/ucb-datahub-2018/primary-r-user-image
```

sidecar reads this config, gets course enrollment and instructors, and writes them to somewhere.d/{year}-{term}-{classid}-students.txt somewhere.d/{year}-{term}-{classid}-instructors.txt. 

And the custom spawner has this pseudo-ish code:
```
  def start(self):
    ...

    is_student = 0 ; is_student = 0 ; set_resources = False
    custom_profiles = z2jh.get_config('custom.profiles', {})
    for course, profile_data in custom_profiles.items():
        course_consituency = read_course_json(course)
        if user_is_course_student(self.user.name, course_consituency):
            is_student += 1
            set_resources = True
        elif user_is_course_instructor(self.user.name, course_consituency):
            is_instructor += 1
            set_resources = True

        if set_resources:
            self.mem_limit = max_memparse(profile_data['mem_limit'], self.mem_limit)
            self.mem_guarantee = max_memparse(profile_data['mem_guarantee'], self.mem_guarantee)

        # if user is a student in any course, they never get to be an admin
        if is_instructor > 0 and is_student == 0:
           add_user_to_jupyterhub_auth_admin_users()
```

**Comment by yuvipanda** [comment]
@ryanlovett this looks great, although I'd like to separate the instructor / student setup into a different issue from this. Does that sound ok?

Also, is there a reason is_student and is_instructor are integers rather than booleans? We only check them against 0, so using a bool should also be fine right?

**Comment by ryanlovett** [comment]
@yuvipanda So keep the resource settings bits in this issue, but move the handling of admin privileges into another?

Totally could use booleans -- would OR the student state since that taints getting admin.

**Comment by yuvipanda** [comment]
@ryanlovett no, more like - we can do the handling of admin privileges for instructors to a different PR later, since I think that has more social implications around who gets admin. Also admin lists are currently public, so I don't see it as a pressing issue to get right.

**Comment by ryanlovett** [comment]
This is performed by a sidecar. We should have another issue that tracks getting enrollment info from bCourses via the CanvasOAuthenticator.
